acceptance/analyzer_test.go

```diff
@@ -0,0 +1,503 @@
+			when("cache image case", func() {
+				it.Before(func() {
+					metadata := flattenMetadata(t, filepath.Join("testdata", "analyzer", "cache_metadata.json"), lifecycle.CacheMetadata{})
+
+					cmd := exec.Command(
+						"docker",
+						"build",
+						"-t", cacheImageRepo,
+						"--build-arg", fmt.Sprintf("metadata=%s", metadata),
+						filepath.Join("testdata", "analyzer", "cache-image"),
+					)
+					h.Run(t, cmd)
+				})
+
+				it.After(func() {
+					removeTestImage(t, cacheImageRepo)
+				})
+
+				it("ignores the cache", func() {
+					cmd := exec.Command(
+						"docker", "run", "--rm",
+						"--mount", "type=bind,source=/var/run/docker.sock,target=/var/run/docker.sock",
+						"--env", "CNB_REGISTRY_AUTH={}", // In practice, we never set this variable in the daemon case. Setting to avoid failure to stat docker config directory when initializing cache.
+						analyzeImage,
+						"/bin/bash", "-c",
+						fmt.Sprintf("%s -daemon -cache-image %s some-image; ls -alR /layers", analyzerPath, cacheImageRepo),
+					)
+					output := h.Run(t, cmd)
+
+					h.AssertNotMatch(t, output, "some-layer.sha")
+					h.AssertNotMatch(t, output, "some-layer.toml")
+				})
+			})
+
+			when("cache directory case", func() {
+				var (
+					cacheDir string
+					err      error
+				)
+
+				it.Before(func() {
+					cacheDir, err = ioutil.TempDir("", "analyzer-acceptance")
+					h.RecursiveCopy(t, filepath.Join("testdata", "analyzer", "cache-dir"), cacheDir)
+					// Some OSes (like macOS) use symlinks for the standard temp dir.
+					// Resolve it so it can be properly mounted by the Docker daemon.
+					cacheDir, err = filepath.EvalSymlinks(cacheDir)
+					h.AssertNil(t, err)
+				})
+
+				it.After(func() {
+					os.RemoveAll(cacheDir)
+				})
+
+				it("uses the provided cache", func() {
+					cmd := exec.Command(
+						"docker", "run", "--rm",
+						"--mount", "type=bind,source=/var/run/docker.sock,target=/var/run/docker.sock",
+						"--volume", fmt.Sprintf("%s:/cache", cacheDir),
+						analyzeImage,
+						"/bin/bash", "-c",
+						fmt.Sprintf("%s -daemon -cache-dir /cache some-image; ls -alR /layers", analyzerPath),
+					)
+					output := h.Run(t, cmd)
+
+					h.AssertMatch(t, output, "/layers/some-buildpack-id")
+					h.AssertMatch(t, output, "some-layer.sha")
+					h.AssertMatch(t, output, "some-layer.toml")
+				})
+
+				when("the provided cache directory isn't writeable", func() {
+					it("recursively chowns the directory", func() {
+						cmd := exec.Command(
+							"docker", "run", "--rm",
+							"--mount", "type=bind,source=/var/run/docker.sock,target=/var/run/docker.sock",
+							"--volume", fmt.Sprintf("%s:/cache", cacheDir),
+							analyzeImage,
+							"/bin/bash", "-c",
+							fmt.Sprintf("%s -daemon -cache-dir /cache some-image; ls -alR /cache", analyzerPath),
+						)
+						output := h.Run(t, cmd)
+
+						h.AssertMatch(t, output, "2222 3333 .+ \\.")
+						h.AssertMatch(t, output, "2222 3333 .+ committed")
```

Comment:
The staging directory is not chowned - is that expected?

Reply:
It looks like it is being created with the wrong permissions:
```
        /cache:
        total 4
        drwxrwxr-x 4 2222 3333  128 Jun 26 19:59 .
        drwxr-xr-x 1 root root 4096 Jun 26 19:59 ..
        drwxrwxr-x 4 2222 3333  128 Aug 30  1754 committed
        drwxr-xr-x 2 root root   64 Jun 26 19:59 staging
```

What's interesting, when I add a staging directory to my fixture, it looks like that is ignored? At least, I don't see the `foo.txt` file that I inserted there. Will investigate...

Reply:
Uggh I was probably doing something wrong - seeing different results now.

Reply:
I still see that foo.txt is ignored, but staging has the correct permissions. I am not sure what changed other than the cache directory is now being provided as a volume vs. bind mount. What is interesting is that in the case where the user's group can write, I see that staging is chowned:

```
/cache:
total 16
drwxrwxr-x 4 9999 3333 4096 Jun 29 20:23 .
drwxr-xr-x 1 root root 4096 Jun 29 20:23 ..
drwxrwxr-x 2 9999 3333 4096 Jun 29 20:21 committed
drwxr-xr-x 2 2222 3333 4096 Jun 29 20:23 staging
```

Adding a note to investigate further. There is some weirdness here that I'm not understanding.

Reply:
Looking at the code, `if cacheDir != ""` we always call `cache.NewVolumeCache(cacheDir)` which always recreates the staging directory (hence `foo.txt` was ignored). This happens after we drop privileges, so it makes sense that the uid & gid are 2222 & 3333. I have no idea why it was showing as owned by root before :/